STYLE_GUIDE.md

```diff
@@ -0,0 +1,156 @@
+ * [golangci (covers all important linters)](https://github.com/golangci/golangci-lint)
+   - See the `.golangcli.yml` file in each repo for linter configuration.
+
+## Various
+
+ * Reserve "Save" and "Load" for long-running persistence operations. When parsing bytes, use "Encode" or "Decode".
+ * Maintain consistency across the codebase.
+ * Functions that return functions should have the suffix `Fn`
+ * A struct generally shouldn’t have a field named after itself, aka. this shouldn't occur:
+``` golang
+type middleware struct {
+	middleware Middleware
+}
+```
+ * In comments, use "iff" to mean, "if and only if".
+ * Product names are capitalized, like "Tendermint", "Basecoin", "Protobuf", etc except in command lines: `tendermint --help`
+ * Acronyms are all capitalized, like "RPC", "gRPC", "API".  "MyID", rather than "MyId".
+ * Prefer errors.New() instead of fmt.Errorf() unless you're actually using the format feature with arguments. (otherwise it's needlessly inefficient, & it'll crap out on %*)
+
+## Importing Libraries
+
+Sometimes it's necessary to rename libraries to avoid naming collisions or ambiguity.
+
+ * Use [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports)
+ * Separate imports into blocks - one for the standard lib, one for external libs and one for tendermint libs.
+ * Here are some common library labels for consistency:
+   - dbm "github.com/tendermint/tm-db"
+   - tmcmd "github.com/tendermint/tendermint/cmd/tendermint/commands"
+   - tmcfg "github.com/tendermint/tendermint/config/tendermint"
+   - tmtypes "github.com/tendermint/tendermint/types"
+ * Never use anonymous imports (the `.`), for example, `tmlibs/common` or anything else.
+ * When importing a pkg from the `tendermint/libs` directory, prefix the pkg alias with tm.
+     - tmbits "github.com/tendermint/tendermint/libs/bits"
+ * tip: Use the `_` library import to import a library for initialization effects (side effects)
+
+## Dependencies
+
+ * Dependencies should be pinned by a release tag, or specific commit, to avoid breaking `go get` when external dependencies are updated.
+
+## Testing
+
+ * The first rule of testing is: we add tests to our code
+ * The second rule of testing is: we add tests to our code
+ * For Golang testing:
+   * Make use of table driven testing where possible and not-cumbersome
+     - [Inspiration](https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go)
+   * Make use of [assert](https://godoc.org/github.com/stretchr/testify/assert) and [require](https://godoc.org/github.com/stretchr/testify/require)
+ * When using mocks, it is recommended to use either [testify mock](https://pkg.go.dev/github.com/stretchr/testify/mock
+ ) or [Mockery](https://github.com/vektra/mockery)
+
+## Errors
+
+ * Ensure that errors are concise, clear and traceable.
+ * Use stdlib errors package.
+ * For wrapping errors, use `fmt.Errorf()` with `%w`.
+ * Use panic sparingly only to expose internal problems.
```

Comment:
Maybe we can directly lift a sentence from this [great explanation](https://github.com/tendermint/tendermint/issues/4465#issuecomment-602705133) from @erikgrinaker here: 

> My recommendation would be that panic is appropriate when an internal invariant of a system is broken, while all other cases (in particular, incorrect or invalid usage) should return errors.

Reply:
Done :+1: